### ISSUE TITLE
Test MAPL v2.53.0 in UFS weather model

### ISSUE THREAD

**Issue by junwang-noaa** [ufs-community/ufs-weather-model#2346]
## Description

MAPL 2.46.2 has fixes for issue #2162. UFS weather model needs to be tested and updated with this version.

20240830:
MAPL 2.46.2 has a bug. MAPL 2.46.3 should be installed and tested in UFS weather model. The issue title is updated.

**Comment by junwang-noaa** [comment]
@lipan-NOAA Can you confirm this version of MAPL works in GOCART for GEFSv13? Thanks

**Comment by BrianCurtis-NOAA** [comment]
@Hang-Lei-NOAA is MAPL 2.46.2 installed on Acorn/WCOSS2?

**Comment by Hang-Lei-NOAA** [comment]
I will add these today.

**Comment by lipan-NOAA** [comment]
@Hang-Lei-NOAA Can you tell me where you installed it?

**Comment by Hang-Lei-NOAA** [comment]
@Li Pan - NOAA Affiliate ***@***.***> Please check here
module use
/lfs/h1/emc/nceplibs/noscrub/hpc-stack/libs/hpc-stack/modulefiles/mpi/intel/19.1.3.304/cray-mpich/8.1.9
module load esmf/8.6.1
module load mapl/4.6.2-esmf-8.6.1

**Comment by ulmononian** [comment]
do you want this installed in spack-stack/1.6.0? and on which machine?

**Comment by jkbk2004** [comment]
@DusanJovic-NOAA @junwang-noaa new maple and esmf version are available on hercules and orion for the test and debug activities. @RatkoVasic-NOAA thanks for the installation!
```
Hercules: /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/ue-esmf-8.6.1-mapl-2.46.3/install/modulefiles/Core
Orion: /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.6.0/envs/ue-esmf-8.6.1-mapl-2.46.3/install/modulefiles/Core
```

**Comment by mathomp4** [comment]
As an FYI, the MAPL 2.46.3 fix was when using externally initialized MPI. (Which is something you all do but we don't do internally).

We also had this notice to users:

> Notice to users: External code should initialize MPI with `MPI_THREAD_MULTIPLE` as that is what MAPL expects. Also, code might need to call:
> ```
> call ESMF_InitializePreMPI()
> ```
> for all features of MAPL to be supported, namely if ESMF Single System Image (SSI) code is enabled. If users do not use this (or know what it is), it most likely is not needed.

My guess is you do not need to call `ESMF_InitializePreMPI()` as I don't think even we internally quite use all the SSI code yet.

**Comment by junwang-noaa** [comment]
@weiyuan-jiang may I ask what compiler/mpi versions you want to test? Thanks

**Comment by mathomp4** [comment]
> @weiyuan-jiang may I ask what compiler/mpi versions you want to test? Thanks

@junwang-noaa Well for that version of MAPL, we would have been testing with:

- Intel ifort 2021.6
- Intel ifort 2021.12 (or .13)
- GCC 13.2.0

We never had Intel ifort 2021.9 on any machines we have. Are any of these possible?

**Comment by junwang-noaa** [comment]
@AlexanderRichert-NOAA do we have these intel and GCC version on Hera or Hercules orGaea?

**Comment by AlexanderRichert-NOAA** [comment]
I can't get onto Hera at the moment but for the others:

- Gaea (C5 & C6):
  - ifort 2021.6 (module load intel-classic/2022.1.0)
  - GCC 13.2 (module load gcc-native/13.2)
- Hercules & Orion:
  - ifort 2021.12.0 (module load intel-oneapi-compilers/2024.1.0)

**Comment by mathomp4** [comment]
As we have access to Hercules/Orion, I guess ifort 2021.12 is our best bet at the moment (though I'd have to imagine there must be a recent-ish gcc on there).

I'm fairly certain we can run with 2021.12. There were some bugfixes needed in places of GEOS (I don't think MAPL, but I can find out), so by the time we got them all in, 2021.13 was out, but I think 2021.12 works.

Plus, we have 2021.12 on discover, so we can do some matching if need be.

**Comment by junwang-noaa** [comment]
@weiyuan-jiang @mathomp4 Which MAPL version would you like us to test in UFS? I assume the ESMF version is 8.6.1

**Comment by weiyuan-jiang** [comment]
@junwang-noaa It doesn't matter. I can always replace the MAPL as long as it can be built and run under new compiler on Hercules

**Comment by junwang-noaa** [comment]
@AlexanderRichert-NOAA would you please build ifort 2021.12.0 with ESMF 8.6.1 and MAPL v2.50.2 on Hercules in spack-stack 1.6.0? Thank you!

**Comment by AlexanderRichert-NOAA** [comment]
Since ifort 2021.12.0 is only available with oneAPI, will it work to use icx/icpx from the same module? If not, I can try mixing and matching to use icc/icpc, but it might take some fiddling with your modulefile to get it to work since the compilers won't all be available in the same module.

**Comment by junwang-noaa** [comment]
@AlexanderRichert-NOAA I actually do not know. Maybe we need to ask Hercules admin.

**Comment by mathomp4** [comment]
> Since ifort 2021.12.0 is only available with oneAPI, will it work to use icx/icpx from the same module? If not, I can try mixing and matching to use icc/icpc, but it might take some fiddling with your modulefile to get it to work since the compilers won't all be available in the same module.

@AlexanderRichert-NOAA MAPL should be good with `icx` and `icpx` at least in recent times from my memory. A few other repos needed updates, but I don't think MAPL did...

**Comment by DusanJovic-NOAA** [comment]
@climbfuji has open PR with changes to support so-called mixed Intel compilers (icx, icpx, ifort) #2504

**Comment by DusanJovic-NOAA** [comment]
Actually, this will be a good test to confirm that we can use a spack_stack build libraries with llvm c/c++ compiler and classic Fortran compiler during the transition period.

**Comment by AlexanderRichert-NOAA** [comment]
I actually ended up running into some difficulties when using icx/icpx (I forget which exact package/s failed), so it ended up being easiest to use icc and icpc with the newer ifort. hercules:/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/modulefiles/Core

**Comment by DusanJovic-NOAA** [comment]
I've updated my branch to use this spack environment, but now compilation fails with errors like this:
```
/work/noaa/fv3-cam/djovic/ufs/update_gocart_mapl/ufs-weather-model/GOCART/Process_Library/GOCART2G_MieMod.F90(12): error #7012: The module file cannot be read.  Its format requires a more recent F90 compiler.   [NETCDF]
  use netcdf
-------^
/work/noaa/fv3-cam/djovic/ufs/update_gocart_mapl/ufs-weather-model/GOCART/Process_Library/GOCART2G_MieMod.F90(182): error #6404: This name does not have a type, and must have an explicit type.   [NF90_NOWRITE]
     rc = nf90_open(this%table_name, NF90_NOWRITE, ncid)
--------------------------------------^
/work/noaa/fv3-cam/djovic/ufs/update_gocart_mapl/ufs-weather-model/GOCART/Process_Library/GOCART2G_MieMod.F90(182): error #6404: This name does not have a type, and must have an explicit type.   [NF90_OPEN]
     rc = nf90_open(this%table_name, NF90_NOWRITE, ncid)
-----------^
```
I'm confused about which version of the compiler this environment is supposed to use. When I list available modules I see:
```
$ module use /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/modulefiles/Core

$ module avail

--------------------------------------- /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/modulefiles/Core ----------------------------------------
   stack-intel/2021.9.0
```
The `stack-intel/2021.9.0` module is:
```
$ module show stack-intel/2021.9.0
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
   /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/modulefiles/Core/stack-intel/2021.9.0.lua:
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
help([[]])
family("MetaCompiler")
conflict("stack-apple-clang")
conflict("stack-clang")
conflict("stack-gnu")
conflict("stack-intel")
conflict("stack-intel-oneapi-compilers")
load("intel-oneapi-compilers/2023.1.0")
prereq("intel-oneapi-compilers/2023.1.0")
prepend_path("MODULEPATH","/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/modulefiles/intel/2021.9.0")
setenv("F77","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2024.1.0-76aou5yu2oqfcz72vq7lp77bp3rqcoht/compiler/2024.1/bin/ifort")
setenv("FC","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2024.1.0-76aou5yu2oqfcz72vq7lp77bp3rqcoht/compiler/2024.1/bin/ifort")
setenv("CC","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/icc")
setenv("CXX","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/icpc")
setenv("SERIAL_F77","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2024.1.0-76aou5yu2oqfcz72vq7lp77bp3rqcoht/compiler/2024.1/bin/ifort")
setenv("SERIAL_FC","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2024.1.0-76aou5yu2oqfcz72vq7lp77bp3rqcoht/compiler/2024.1/bin/ifort")
setenv("SERIAL_CC","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/icc")
setenv("SERIAL_CXX","/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/icpc")
setenv("I_MPI_EXTRA_FILESYSTEM","ON")
whatis("Name: stack-intel")
whatis("Version: 2021.9.0")
whatis("Category: compiler")
whatis("Description: stack-intel compiler family and module access")
```

When I load it:
```
$ module load stack-intel/2021.9.0

$ echo $FC
/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2024.1.0-76aou5yu2oqfcz72vq7lp77bp3rqcoht/compiler/2024.1/bin/ifort

$ which ifort
/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/ifort
```
It sets $FC environment variable to ifort 2024.1, but since it also loads `intel-oneapi-compilers/2023.1.0` the ifort in the $PATH is 2023.1. 

Module `stack-intel` version `2021.9.0` sets FC variable to the compiler version `2024.1.0` but also adds path to compiler version `2023.1.0`. Very confusing, and inconsistent.

Is there anything I need to change in the ufs-weather-model modulefiles in order to fix this? Set some variable, remove or add something?

**Comment by AlexanderRichert-NOAA** [comment]
Can you run make verbosely to verify which ifort it's using in that step?

**Comment by DusanJovic-NOAA** [comment]
```
[  0%] Building Fortran object GOCART/Process_Library/CMakeFiles/Process_Library.dir/GOCART2G_MieMod.F90.o
cd /work2/noaa/stmp/djovic/stmp/djovic/FV3_RT/rt_751888/compile_s2swa_intel/build_fv3_s2swa_intel/GOCART/Process_Library && /apps/spack-managed/oneapi-2023.1.0/intel-oneapi-mpi-2021.9.0-a66eaipzsnyrdgaqzxmqmqz64qzvhkse/mpi/2021.9.0/bin/mpiifort -DESMA64 -DHAS_NETCDF3 -DsysLinux -I/work/noaa/fv3-cam/djovic/ufs/update_gocart_mapl/ufs-weather-model/GOCART/Process_Library -I/work2/noaa/stmp/djovic/stmp/djovic/FV3_RT/rt_751888/compile_s2swa_intel/build_fv3_s2swa_intel/GOCART/Process_Library -I/work2/noaa/stmp/djovic/stmp/djovic/FV3_RT/rt_751888/compile_s2swa_intel/build_fv3_s2swa_intel/include/Process_Library -I/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/intel/2021.9.0/esmf-8.6.1-7z7d4jp/include -I/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/intel/2021.9.0/netcdf-c-4.9.2-tttotzi/include -I/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/intel/2021.9.0/netcdf-fortran-4.6.1-4qjzskp/include -I/work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/intel/2021.9.0/parallelio-2.5.10-b7pdfd6/include -g -traceback -fpp -fno-alias -auto -safe-cray-ptr -ftz -assume byterecl -sox -align array64byte -qno-opt-dynamic-align -diag-disable 5462 -diag-disable 7712 -diag-disable 8291  -O2 -debug minimal -qoverride-limits -fp-model consistent -march=core-avx2 -module ../../include/Process_Library -qopenmp -c /work/noaa/fv3-cam/djovic/ufs/update_gocart_mapl/ufs-weather-model/GOCART/Process_Library/GOCART2G_MieMod.F90 -o CMakeFiles/Process_Library.dir/GOCART2G_MieMod.F90.o
```

**Comment by AlexanderRichert-NOAA** [comment]
Can you try setting I_MPI_CC/I_MPI_FC to the appropriate compiler paths in the modulefile?

**Comment by DusanJovic-NOAA** [comment]
I do not know which are the appropriate compiler paths. How do I find that? 

When I do:

```
$ module use /work/noaa/epic/role-epic/spack-stack/hercules/spack-stack-1.6.0/envs/esmf-8.6.1-mapl-2.50.2-intel-mixed/install/modulefiles/Core

$ module load stack-intel/2021.9.0

$ ml

Currently Loaded Modules:
  1) intel-oneapi-compilers/2023.1.0   2) stack-intel/2021.9.0
```
I see these paths:

```
$ which icc
/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/icc

$ which icpc
/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/icpc

$ which ifort
/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2023.1.0-sb753366rvywq75zeg4ml5k5c72xgj72/compiler/2023.1.0/linux/bin/intel64/ifort
```

**Comment by AlexanderRichert-NOAA** [comment]
The goal here is to try to use icc from the 2023.1.0 installation with ifort from 2024.1; we're loading the 2023.1.0 module, so I _believe_ what we want to do is set I_MPI_FC=/apps/spack-managed/gcc-11.3.1/intel-oneapi-compilers-2024.1.0-76aou5yu2oqfcz72vq7lp77bp3rqcoht/compiler/**2024.1**/bin/ifort so that it will use that one rather than searching $PATH and using the older ifort.

**Comment by DusanJovic-NOAA** [comment]
I updated this branch https://github.com/DusanJovic-NOAA/ufs-weather-model/tree/update_gocart_mapl to use the icc/icpc from 2031.1.0 and ifort from 2024.1.0.

cpld_control_p8 test still fails. Same errors as before:

```
122: pe=00122 FAIL at line=02772    Base_Base_implementation.F90             <It only works for cubed-sphere grid>
122: pe=00122 FAIL at line=02634    Base_Base_implementation.F90             <status=1>
122: pe=00122 FAIL at line=00695    SU2G_GridCompMod.F90                     <status=1>
```

**Comment by lipan-NOAA** [comment]
I tested the cpld_bmark_p8_intel case and received the same error message as Dusan.
**GOCART fails when reading sulfate point emissions**
":SU::SU::SU::SU::Run1 - cannot get indices for point emissions"

**Comment by weiyuan-jiang** [comment]
I am not sure how it is upgraded. It is still 2021.9 to me.

git pull
module use modulefiles

module load ufs_hercules.intel
icc --version
icc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
icc (ICC) 2021.9.0 20230302

ifort --version
ifort (IFORT) 2021.9.0 20230302

**Comment by DusanJovic-NOAA** [comment]
@weiyuan-jiang I repeated 'cpld_control_p8 intel' test and confirmed that version 2021.12.0 is used (see: /work2/noaa/stmp/djovic/stmp/djovic/FV3_RT/rt_1593997/cpld_control_p8_intel/out). ifort command is still showing 2012.9.0 but in the module file we set I_MPI_F90:

https://github.com/DusanJovic-NOAA/ufs-weather-model/blob/9296c85645444541558f052d1a48717ead654475/modulefiles/ufs_hercules.intel.lua#L24

so when mpiifort wrapper is used, the version of the compiler used is:

```
$ mpiifort --version
ifort: remark #10448: Intel(R) Fortran Compiler Classic (ifort) is now deprecated and will be discontinued late 2024. Intel recommends that customers transition now to using the LLVM-based Intel(R) Fortran Compiler (ifx) for continued Windows* and Linux* support, new language support, new language features, and optimizations. Use '-diag-disable=10448' to disable this message.
ifort (IFORT) 2021.12.0 20240222
Copyright (C) 1985-2024 Intel Corporation.  All rights reserved.
```

Alex was not able to use both C and Fortran compiler from 2021.12.0 to build libraries, that's why this hack was necessary, see above comments.

Also, take a look at the err file from my last run: /work2/noaa/stmp/djovic/stmp/djovic/FV3_RT/rt_1593997/cpld_control_p8_intel/err

**Comment by weiyuan-jiang** [comment]
@DusanJovic-NOAA Thanks.  Is the ESMF library also build by the same compiler ?  I think ESMF should be built by 2021.12 because it is ESMF's call that crashes

**Comment by DusanJovic-NOAA** [comment]
I think it is. @AlexanderRichert-NOAA can you confirm (check?) that esmf is compiled with fortran version 2012.12.0 Thanks.

**Comment by weiyuan-jiang** [comment]
It seems to me the ipc and icpc is still 2021.9.  ESMF are written in C and C++ mostly

**Comment by DusanJovic-NOAA** [comment]
According to this comment https://github.com/ufs-community/ufs-weather-model/issues/2346#issuecomment-2524030585, 2021.12.0 has only icx/icpx and ifort, and there were some issues using icx, if I understand correctly.

**Comment by weiyuan-jiang** [comment]
If it is mixed,  can it go back to the older ipc and icps like 2021.6 ? We don't have 2021.9 in our system so we cannot reproduce the issue.

**Comment by DavidHuber-NOAA** [comment]
FWIW the only NOAA system I see with 2021.6 is Gaea (C5 and C6) via the intel/2022.1.0 module.

**Comment by DusanJovic-NOAA** [comment]
I see this esmf error in PET* log files when I run `atmaero_control_p8` test:

```
20241223 080207.562 INFO             PET000 UFS Aerosols: Advancing from 2021-03-22T06:00:00 to 2021-03-22T06:12:00
20241223 080219.317 ERROR            PET000 ESMCI_Info.C:668 Info::erase() Not found  - [json.exception.out_of_range.403] key 'GridCornerLons:' not found
20241223 080219.317 ERROR            PET000 ESMCI_Info.C:688 Info::erase() Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 ESMC_InfoCDef.C:243 ESMC_InfoErase() Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 ESMF_Info.F90:2661 ESMF_InfoRemove() Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 src/Superstructure/AttributeAPI/interface/ESMF_Attribute.F90:46022 ESMF_AttributeRemoveAttPackGrid( Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 ESMCI_Info.C:668 Info::erase() Not found  - [json.exception.out_of_range.403] key 'GridCornerLats:' not found
20241223 080219.318 ERROR            PET000 ESMCI_Info.C:688 Info::erase() Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 ESMC_InfoCDef.C:243 ESMC_InfoErase() Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 ESMF_Info.F90:2661 ESMF_InfoRemove() Not found  - Internal subroutine call returned Error
20241223 080219.318 ERROR            PET000 src/Superstructure/AttributeAPI/interface/ESMF_Attribute.F90:46022 ESMF_AttributeRemoveAttPackGrid( Not found  - Internal subroutine call returned Error
```
Could this be the reason for the error we see in mapl where the grid is not recognized as cubed-sphere grid?

**Comment by DeniseWorthen** [comment]
This looks to be the same message I saw in https://github.com/ufs-community/ufs-weather-model/issues/1888

**Comment by natalie-perlin** [comment]
The message "json exception" sounds a bit familiar... (https://github.com/ufs-community/ufs-weather-model/discussions/2371#discussioncomment-11018632)

What helped to resolve this issue on MacOS (clang for C, CXX, and gfortran for Fortran) was using CXX linker instead of Fortran. The issue could arise when mixing C/CXX compilers and Fortran compilers from different vendors.
https://github.com/ufs-community/ufs-weather-model/discussions/2371#discussioncomment-11123529
The change is to use CXX instead of Fortran in CMakeLists.txt

`set_target_properties(ufs_model PROPERTIES LINKER_LANGUAGE CXX)`

This may not be the solution for this particular case, but it may worth testing as an option.
(my 2c)

**Comment by DusanJovic-NOAA** [comment]
> The message "json exception" sounds a bit familiar... ([#2371 (comment)](https://github.com/ufs-community/ufs-weather-model/discussions/2371#discussioncomment-11018632))
> 
> What helped to resolve this issue on MacOS (clang for C, CXX, and gfortran for Fortran) was using CXX linker instead of Fortran. The issue could arise when mixing C/CXX compilers and Fortran compilers from different vendors. [#2371 (comment)](https://github.com/ufs-community/ufs-weather-model/discussions/2371#discussioncomment-11123529) The change is to use CXX instead of Fortran in CMakeLists.txt
> 
> `set_target_properties(ufs_model PROPERTIES LINKER_LANGUAGE CXX)`
> 
> This may not be the solution for this particular case, but it may worth testing as an option. (my 2c)

I tried setting LINKER_LANGUAGE to CXX but compilation fails:
```
of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
icpc: command line warning #10006: ignoring unknown option '-threads'
ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/crt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
make[2]: *** [CMakeFiles/ufs_model.dir/build.make:153: ufs_model] Error 1
make[1]: *** [CMakeFiles/Makefile2:642: CMakeFiles/ufs_model.dir/all] Error 2
make: *** [Makefile:136: all] Error 2                                                                                                  
```

**Comment by natalie-perlin** [comment]
> > The message "json exception" sounds a bit familiar... ([#2371 (comment)](https://github.com/ufs-community/ufs-weather-model/discussions/2371#discussioncomment-11018632))
> > What helped to resolve this issue on MacOS (clang for C, CXX, and gfortran for Fortran) was using CXX linker instead of Fortran. The issue could arise when mixing C/CXX compilers and Fortran compilers from different vendors. [#2371 (comment)](https://github.com/ufs-community/ufs-weather-model/discussions/2371#discussioncomment-11123529) The change is to use CXX instead of Fortran in CMakeLists.txt
> > `set_target_properties(ufs_model PROPERTIES LINKER_LANGUAGE CXX)`
> > This may not be the solution for this particular case, but it may worth testing as an option. (my 2c)
> 
> I tried setting LINKER_LANGUAGE to CXX but compilation fails:
> 
> ```
> of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
> icpc: command line warning #10006: ignoring unknown option '-threads'
> ld: /usr/lib/gcc/x86_64-redhat-linux/11/../../../../lib64/crt1.o: in function `_start':
> (.text+0x1b): undefined reference to `main'
> make[2]: *** [CMakeFiles/ufs_model.dir/build.make:153: ufs_model] Error 1
> make[1]: *** [CMakeFiles/Makefile2:642: CMakeFiles/ufs_model.dir/all] Error 2
> make: *** [Makefile:136: all] Error 2                                                                                                  
> ```

Thank you very much for testing this option, good to know that this potential quick fix had other implications... I will look more into this issue with the "json.exception error". It has something to do with nlohmann/json library in C++, which may need help to be located or linked.

**Comment by weiyuan-jiang** [comment]
When the error said "undefined reference to `main'", it usually means the link language should be Fortran.

**Comment by DusanJovic-NOAA** [comment]
I think I found the reason why model reports the error message "<It only works for cubed-sphere grid>" and ultimately fails.

```
122: pe=00122 FAIL at line=02772    Base_Base_implementation.F90             <It only works for cubed-sphere grid>
122: pe=00122 FAIL at line=02634    Base_Base_implementation.F90             <status=1>
122: pe=00122 FAIL at line=00695    SU2G_GridCompMod.F90                     <status=1>

```
The above error message is first printed in "MAPL_GetGlobalHorzIJIndex" subroutine called from "MAPL_GetHorzIJIndex" which is called from "Run1" routine in SU2G_GridCompMod. The reason the error is printed is that `IM_WORLD*6 != JM_WORLD`. IM_WORLD, JM_WORLD are returned from MAPL_GridGet as globalCellCountPerDim where JM_WORLD is set to be JM_WORLD*6 only if tileCount is 6, which it should be for "cubed-sphere" grids. The grid is passed as subroutine argument from Run1 is SU2G_GridCompMod. I printed the value of tileCount in Run1, and it is 1, not 6. When I print tileCount of the grid in 'Initialize' phase of the SU2G grid component it is 6, which is consistent with the grid created in fv3atm and which is the grid used in gocart. In "Initialize" phase the grid is retrieved from the grid component:

```
   call ESMF_GridCompGet (GC, grid=grid, name=COMP_NAME, config=universal_cfg, __RC__)
```
while in the "Run1" phase the grid is retrieved from the mapl object:
```
    call ESMF_GridCompGet (GC, NAME=COMP_NAME, __RC__)
    call MAPL_GetObjectFromGC (GC, mapl, __RC__)
    call MAPL_Get(mapl, grid=grid, __RC__)
```
The reason grids in "initialize" and "run1" are different is because the grid from the mapl object in 'run1' is actually a subgrid computed from the original 'cubed-sphere' grid passed from fv3atm in case gocart/mapl are running using omp threading. The subgrids created for each of the omp threads are not 'cubed-sphere' grids (with tileCount == 6), instead they are just regular single tile rectangular grids. This is done in 'make_subgrids_from_bounds' in generic/OpenMP_Support.F90 [here](https://github.com/GEOS-ESM/MAPL/blob/main/generic/OpenMP_Support.F90#L113-L122).

I think the code that creates the subgrids should be updated to correctly create multi-tiled grids if the original (primary_grid) is multi-tiled. Or maybe the logic in MAPL_GridGet should be updated to not rely on the value of tileCount to compute globalCellCountPerDim.

As a quick fix, if I run the test without omp threads in gocart by setting "use_threads: .FALSE." it does not print those "It only works for cubed-sphere grid" FAIL messages. 

Unfortunately it still fails with floating point exception later in the forecast.

**Comment by weiyuan-jiang** [comment]
@DusanJovic-NOAA Thanks for finding this. At this point I can confirm that multi threading in MAPL ( except the older one like 2.40...)  would not work with GOCART.

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA Thanks for finding this. At this point I can confirm that multi threading in MAPL ( except the older one like 2.40...) would not work with GOCART.

Thanks for confirming.

**Comment by DusanJovic-NOAA** [comment]
The model is failing with the floating point exception because of grid mask inconsistencies between fv3atm/gocart and what MAPL expects. MAPL is computing remapping route handles in "create_route_handle" in base/MAPL_EsmfRegridder.F90 and it sets 'dstMaskValues' to MAPL_MASK_OUT (which is 0):
```
          if (has_mask) dstMaskValues = [MAPL_MASK_OUT] ! otherwise unallocated
```
where "has_mask" is true if grid has ESMF_GRIDITEM_MASK item. The fv3grid has ESMF_GRIDITEM_MASK set, and the mask values are either 0=ocean or 1=land. Which means the grid points with mask = 0 are valid destination points and should not be masked out. Currently, by using hard-coded value MAPL_MASK_OUT constant it is assumed that all mask points with 0 are masked out, which is currently not the case in fv3atm and gocart when used with fv3atm. Maybe value can be passed via grid attributes or something like that. Or maybe in fv3atm we can change the mask values to something other than 0, for example, 1=land, 2=ocean, but I do not know if and how that will impact other coupled components and CMEPS.

Anyway, as a test I simply commented the above line that sets the value of dstMaskValues, and the model finally finished the test without any error.

**Comment by DeniseWorthen** [comment]
@DusanJovic-NOAA I do not think we want to change the current (0,1) mask values for ATM because that will have downstream impacts in CMEPS. 

I can't even find the MAPL code you reference, but if you don't want to mask the destination anywhere, then why not provide a special value? CMEPS uses ``ispval_mask     = -987987 ! spval for RH mask values``

**Comment by DusanJovic-NOAA** [comment]
> @DusanJovic-NOAA I do not think we want to change the current (0,1) mask values for ATM because that will have downstream impacts in CMEPS.
> 
> I can't even find the MAPL code you reference, but if you don't want to mask the destination anywhere, then why not provide a special value? CMEPS uses `ispval_mask = -987987 ! spval for RH mask values`

MAPL code I reference is here:
https://github.com/GEOS-ESM/MAPL/blob/1c55973c244834ed0d78c39964d2738c3d8f8a8b/base/MAPL_EsmfRegridder.F90#L1519

I do not understand where that special value should be provided and how is MAPL going to use that value?

**Comment by DeniseWorthen** [comment]
I didn't mean that MAPL should somehow reach into CMEPS for the value. I was just pointing out that by setting a special value (one that is never encountered), you can map to all destination points.

**Comment by DusanJovic-NOAA** [comment]
> I didn't mean that MAPL should somehow reach into CMEPS for the value. I was just pointing out that by setting a special value (one that is never encountered), you can map to all destination points.

Sorry, I still do not understand what should be set to a special value, a grid mask? The problem is that some of the fv3atm grid points have grid mask set to 0 and MAPL considers them as destination grid points that should be masked out. 

The fv3atm grid mask is defined in this routine https://github.com/NOAA-EMC/fv3atm/blob/a7d46eee01a78f0915373ebc58c9b20ba14a6c36/atmos_model.F90#L3654

**Comment by DeniseWorthen** [comment]
The ``dstMaskValue`` can be set to a special value. 

I'm probably misunderstanding, since I've never looked at MAPL. But the issue is that when the destination is ATM, you want all destination points mapped, right? Or is the destination in this case the aerosol "grid/mesh" ?

**Comment by DusanJovic-NOAA** [comment]
> The `dstMaskValue` can be set to a special value.

Yes, it can be. But it is currently set to 0 (MAPL_MASK_OUT is 0). [Here](https://github.com/GEOS-ESM/MAPL/blob/1c55973c244834ed0d78c39964d2738c3d8f8a8b/base/MAPL_EsmfRegridder.F90#L1519). That means any fv3atm destination point with grid mask 0 (all ocean points) will be masked out.

> 
> I'm probably misunderstanding, since I've never looked at MAPL. But the issue is that when the destination is ATM, you want all destination points mapped, right? Or is the destination in this case the aerosol "grid/mesh" ?

Yes. All fv3atm destination points should be mapped, ie. it should not have grid mask set to 0.

If we can not redefine fv3atm grid mask to not use 0 for any grid point, which it seems we can not do, then either we pass the information to MAPL to not use 0 to mask out destination points. Or maybe somewhere in fv3atm or maybe in gocart we redefine grid mask that is going to be passed to MAPL.

**Comment by DeniseWorthen** [comment]
But why does the ``dstMaskValue`` have to be set using the values that ATM uses to define it's mask? I realize that may be how the code is working, but I don't think it should be constrained like that.

**Comment by junwang-noaa** [comment]
@weiyuan-jiang @tclune UFS is using a land sea mask on FV3 cubed sphere grid (0 for ocean and 1 for land) for coupling purposes. May I ask if the following line (MAPL_MASK_OUT) can be updated so that mask value 0 in UFS won't be considered undefined? 
```
          if (has_mask) dstMaskValues = [MAPL_MASK_OUT] ! otherwise unallocated
```

**Comment by weiyuan-jiang** [comment]
I think the constant value of MAPL_MASK_OUT should be updated

**Comment by DusanJovic-NOAA** [comment]
With this change in GOCART I was able to make a successful (no floating point exception) run.

```diff
diff --git a/ESMF/UFS/Aerosol_Cap.F90 b/ESMF/UFS/Aerosol_Cap.F90
index b753afa..e21ed11 100644
--- a/ESMF/UFS/Aerosol_Cap.F90
+++ b/ESMF/UFS/Aerosol_Cap.F90
@@ -243,6 +243,7 @@ contains
     type(MAPL_CapOptions)     :: maplCapOptions
     type(Aerosol_InternalState_T) :: is
     type(Aerosol_Tracer_T), pointer :: trp
+    integer(kind=ESMF_KIND_I4), pointer  :: maskPtr(:,:)
 
     ! begin
     rc = ESMF_SUCCESS
@@ -338,6 +339,10 @@ contains
       file=__FILE__)) &
       return  ! bail out
 
+    call ESMF_GridGetItem(grid, itemflag=ESMF_GRIDITEM_MASK,   &
+                          staggerloc=ESMF_STAGGERLOC_CENTER, farrayPtr=maskPtr, _RC)
+    maskPtr = 1
+
     ! provide model grid to MAPL
     call cap % cap_gc % set_grid(grid, lm=nlev, _RC)
```
This change explicitly sets all grid mask points to 1, before passing the grid to MAPL.

**Comment by weiyuan-jiang** [comment]
Ben has a new PR (https://github.com/GEOS-ESM/MAPL/pull/3281) to solve the problem on MAPL. So no need to change on ufs

**Comment by DusanJovic-NOAA** [comment]
> Ben has a new PR ([GEOS-ESM/MAPL#3281](https://github.com/GEOS-ESM/MAPL/pull/3281)) to solve the problem on MAPL. So no need to change on ufs

@weiyuan-jiang Thanks for the update. It's good to know that no changes in fv3atm nor gocart are necessary to solve the issues with masking in MAPL regridding.

Any progress with cubed sphere subgrids issue when threading is used.

**Comment by weiyuan-jiang** [comment]
@DusanJovic-NOAA I have opened an issue on MAPL. (https://github.com/GEOS-ESM/MAPL/issues/3274) .  It seems multi-threading is not working at this point

**Comment by DusanJovic-NOAA** [comment]
@weiyuan-jiang I see the MAPL team has released version 2.52.0, which includes the fix for grid masks. I tested it together with recently releases ESMF 8.8.0 on Hercules and all ufs-weather-model regression test passed (using intel compiler, I did not test gnu). I had to turn off OpenMP threading in GOCART due to the first issue with cubed-sphere grids and threading.

**Comment by climbfuji** [comment]
@DusanJovic-NOAA Do you want mapl to be updated to 2.52.0 for spack-stack-1.9.0 (we already have esmf 8.8.0)? We are hoping to cut the release branch tomorrow, thus this would be just in time. Thanks!

**Comment by mathomp4** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/48258889?v=4&amp;size=80">@DusanJovic-NOAA](https://github.com/DusanJovic-NOAA?rgh-link-date=2025-01-23T17%3A54%3A13.000Z) Do you want mapl to be updated to 2.52.0 for spack-stack-1.9.0 (we already have esmf 8.8.0)? We are hoping to cut the release branch tomorrow, thus this would be just in time. Thanks!

Note: We do think we have the fix for the GOCART OpenMP issue here:

https://github.com/GEOS-ESM/MAPL/pull/3350

I believe it worked for @weiyuan-jiang on Hercules, but if it works for you, I could get a MAPL version out tomorrow.

Also, MAPL 2.52 doesn't (yet) require ESMF 8.8 as we don't need any features from that yet in MAPL2. MAPL3 does currently require it.

**Comment by DusanJovic-NOAA** [comment]
> > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/48258889?v=4&amp;size=80">@DusanJovic-NOAA](https://github.com/DusanJovic-NOAA?rgh-link-date=2025-01-23T17%3A54%3A13.000Z) Do you want mapl to be updated to 2.52.0 for spack-stack-1.9.0 (we already have esmf 8.8.0)? We are hoping to cut the release branch tomorrow, thus this would be just in time. Thanks!
> 
> Note: We do think we have the fix for the GOCART OpenMP issue here:
> 
> [GEOS-ESM/MAPL#3350](https://github.com/GEOS-ESM/MAPL/pull/3350)
> 
> I believe it worked for [@weiyuan-jiang](https://github.com/weiyuan-jiang) on Hercules, but if it works for you, I could get a MAPL version out tomorrow.
> 
> Also, MAPL 2.52 doesn't (yet) require ESMF 8.8 as we don't need any features from that yet in MAPL2. MAPL3 does currently require it.

Thanks for the update @mathomp4. I am using a shell script provided by @AlexanderRichert-NOAA  to build chained spack environment with updated ESMF and MAPL to run test on Hercules. If you can make a temporary tag or branch or something that spack can point to (I'm not sure if it has to be release), I should be able to try that version that includes OpenMP fix from 3350 PR you mentioned above. And if we can get that tested before @climbfuji needs to cut release for 1.9.0 spack-stack that would be perfect.

**Comment by climbfuji** [comment]
Other components/models need esmf 8.8.0, might as well use that across the board if it works for Dusan. Thanks Matt!

**Comment by mathomp4** [comment]
@DusanJovic-NOAA I've tagged with `temp-tag-for-dusan` if that's easy to use. It's also at commit https://github.com/GEOS-ESM/MAPL/pull/3350/commits/260d563d08ff0d3e8436ec33ed98a362110e731b

**Comment by DusanJovic-NOAA** [comment]
> [@DusanJovic-NOAA](https://github.com/DusanJovic-NOAA) I've tagged with `temp-tag-for-dusan` if that's easy to use. It's also at commit [GEOS-ESM/MAPL@260d563](https://github.com/GEOS-ESM/MAPL/commit/260d563d08ff0d3e8436ec33ed98a362110e731b)

Thanks. @AlexanderRichert-NOAA is there a way to use this tag (specific commit hash) in your env chainer script to build MAPL.

**Comment by AlexanderRichert-NOAA** [comment]
> Thanks. [@AlexanderRichert-NOAA](https://github.com/AlexanderRichert-NOAA) is there a way to use this tag (specific commit hash) in your env chainer script to build MAPL.

@DusanJovic-NOAA in theory yes-- I just made a small tweak in the env chainer script to allow it to use Spack's built-in support for using git references. For the mapl spec, it would be something like `mapl@git.260d563d08ff0d3e8436ec33ed98a362110e731b=develop` where 'develop' is the version that Spack is going to _treat it as_ (you can set this to a numbered version for the purpose of determining dependencies etc.). The only trick is that in that commit, MAPL's CMakeLists.txt requires CMake 3.24 or higher, which is not what spack-stack-1.6.0 uses, so you have to work around that. `mapl@git.260d563d08ff0d3e8436ec33ed98a362110e731b=develop cmake@3.24.4` for the specs worked for me on hercules just now (with another 'git pull' of the env chainer script, and not forgetting about the `depends_on("udunits")`).

**Comment by DusanJovic-NOAA** [comment]
> > Thanks. [@AlexanderRichert-NOAA](https://github.com/AlexanderRichert-NOAA) is there a way to use this tag (specific commit hash) in your env chainer script to build MAPL.
> 
> [@DusanJovic-NOAA](https://github.com/DusanJovic-NOAA) in theory yes-- I just made a small tweak in the env chainer script to allow it to use Spack's built-in support for using git references. For the mapl spec, it would be something like `mapl@git.260d563d08ff0d3e8436ec33ed98a362110e731b=develop` where 'develop' is the version that Spack is going to _treat it as_ (you can set this to a numbered version for the purpose of determining dependencies etc.). The only trick is that in that commit, MAPL's CMakeLists.txt requires CMake 3.24 or higher, which is not what spack-stack-1.6.0 uses, so you have to work around that. `mapl@git.260d563d08ff0d3e8436ec33ed98a362110e731b=develop cmake@3.24.4` for the specs worked for me on hercules just now (with another 'git pull' of the env chainer script, and not forgetting about the `depends_on("udunits")`).

Thanks. Let me try that.

**Comment by mathomp4** [comment]
I just confirmed with @weiyuan-jiang that he didn't do anything else after using that new MAPL code. He kept OpenMP on in his test.

He did mention you'll need to make sure you have `udunits` loaded, but @AlexanderRichert-NOAA indicated that as well.

**Comment by DusanJovic-NOAA** [comment]
I successfully compiled MAPL tag (commit 260d563d08ff0d3e8436ec33ed98a362110e731b) using @AlexanderRichert-NOAA instructions, and I am running all intel ufs-wm regression tests. So far all tests passed (most importantly, cpld_control_p8_intel passed, which is the test that was failing previously). This is the configuration with:
```
$ grep use_threads parm/gocart/GOCART2G_GridComp.rc
use_threads: .TRUE.
```
So looks like we finally have a version of MAPL that fixes both issues, cubed_sphere grid with threading and masking of ESMF_grid due to mask value conflict between fv3atm and mapl.

In the meantime I will try to build spack-stack using GNU, and run gnu ufs-wm regression tests, just to be sure.

**Comment by DusanJovic-NOAA** [comment]
I should also mention that in order to use ufs-wm with ESMF v8.8.0 an update in GOCART was needed. In my tests, I used a branch @theurich created that has some updates for esmf 8.8.0. [This one](https://github.com/theurich/GOCART/tree/feature/cap-semantic-labels). So before we can actually switch to 8.8.0 this change will need to be merged in gocart. The only issue is that we (ufs-wm) are still using a gocart commit from Sep 2023. And Gerhard's changes are on top of that commit. Somebody will need to port Gerhard's changes to newer gocart and then update ufs-wm to use that version.

**Comment by climbfuji** [comment]
@mathomp4 or @DusanJovic-NOAA Would you mind creating an issue in github.com/jcsda/spack-stack to update MAPL to version x.y.z for spack-stack-1.9.0, and list changes to the build options (if applicable)? Thanks!

**Comment by mathomp4** [comment]
I'm going to work on getting MAPL 2.53 out now. Shouldn't take me too long. Then you'll have a semver tag rather than a crazy git-hash.

ETA: Well, might take about 30-40 minutes. My final test is a full run on discover to make sure it is zero-diff, etc. and good ol' CMake has decided to rebuild the full model. I really hope `ifx` doesn't have this same CMake+`ifort` issue with recompiling!

**Comment by mathomp4** [comment]
MAPL 2.53.0 has been released:

https://github.com/GEOS-ESM/MAPL/releases/tag/v2.53.0

And I have a PR to spack:

https://github.com/spack/spack/pull/48712

and an issue for JCSDA/spack-stack:

https://github.com/JCSDA/spack-stack/issues/1469

Though that last one might not be "complete" as I'm not sure if any updates are needed for how MAPL is built, etc.

**Comment by climbfuji** [comment]
Thanks @mathomp4 !

**Comment by Hang-Lei-NOAA** [comment]
The mapl/2.53.0 requires the cmake min_version as 3.24.0, which is not on
wcoss2.

**Comment by BrianCurtis-NOAA** [comment]
> The mapl/2.53.0 requires the cmake min_version as 3.24.0, which is not on
> wcoss2.
> […](#)

Can you let me know if you install that cmake version on Acorn so I can test it with the current develop version of the UFSWM.

**Comment by Hang-Lei-NOAA** [comment]
This is system lib, we cannot install it. It has to go through system
management.

**Comment by BrianCurtis-NOAA** [comment]
> This is system lib, we cannot install it. It has to go through system
> management.
> […](#)

OK, then we should get the process started to install it.

**Comment by Hang-Lei-NOAA** [comment]
I have modified the CMakelist in mapl/2.53.0, by setting cmake min version
to 2.20.0. The build is successful.

module use
/lfs/h2/emc/eib/save/hang.lei/forgdit/nco_wcoss2/install2/modulefiles/mpi/intel/19.1.3.304/cray-mpich/8.1.12
module load esmf/8.8.0
module show mapl/2.53.0-esmf-8.8.0
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 /lfs/h2/emc/eib/save/hang.lei/forgdit/nco_wcoss2/install2/modulefiles/mpi/intel/19.1.3.304/cray-mpich/8.1.12/mapl/2.53.0-esmf-8.8.0.lua:
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
help([[]])
conflict("mapl")
setenv("MAPL_ROOT","/lfs/h2/emc/eib/save/hang.lei/forgdit/nco_wcoss2/install2/intel-19.1.3.304/cray-mpich-8.1.12/mapl/2.53.0-esmf-8.8.0")
whatis("Name: mapl")
whatis("Version: 2.53.0-esmf-8.8.0")
whatis("Category: library")
whatis("Description: MAPL is a foundation layer of the GEOS architecture")

**Comment by mathomp4** [comment]
> I have modified the CMakelist in mapl/2.53.0, by setting cmake min version to 2.20.0. The build is successful.

Ugh. I guess I figured CMake 3.24 was not new enough to cause issues as it came out in 2022! And, I guess, that maybe spack-stack folks would be building cmake via spack.

I moved MAPL to 3.24 because we needed it for Python support. That was the oldest version of CMake I could get to work with Python 3.12 f2py which moved to meson. Plus Python `FIND_STRATEGY LOCATION` also seemed to need 3.24 in testing.

That said, I have no idea if MAPL can use CMake 2. That is *ANCIENT*. Did you mean CMake 3.20?

**Comment by Hang-Lei-NOAA** [comment]
Yes, I used 2.20.3 installed on wcoss2

**Comment by mathomp4** [comment]
> Yes, I used 2.20.3 installed on wcoss2

That seems...weird. I'm not even sure there *was* a 2.20.3. I mean, there was a CMake 2.8.12.2 in 2018, but I'm not sure the Kitware folks had more in the 2-series before CMake 3.0.0 came out.

**Comment by DavidHuber-NOAA** [comment]
Looking on Cactus, the system version of CMake is 3.20.4.  There are also modules for 3.20.2 and 3.18.4.

**Comment by Hang-Lei-NOAA** [comment]
Yes , I mean that  I set 3.20 up in the CMAkelist of mapl/2.53.0

**Comment by mathomp4** [comment]
Ah. Okay. That makes more sense. Still old, but not quite as old. :)

**Comment by DusanJovic-NOAA** [comment]
> I have modified the CMakelist in mapl/2.53.0, by setting cmake min version
> to 2.20.0. The build is successful.
> 
> module use
> /lfs/h2/emc/eib/save/hang.lei/forgdit/nco_wcoss2/install2/modulefiles/mpi/intel/19.1.3.304/cray-mpich/8.1.12
> module load esmf/8.8.0
> module show mapl/2.53.0-esmf-8.8.0
> -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> 
>  /lfs/h2/emc/eib/save/hang.lei/forgdit/nco_wcoss2/install2/modulefiles/mpi/intel/19.1.3.304/cray-mpich/8.1.12/mapl/2.53.0-esmf-8.8.0.lua:
> -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> help([[]])
> conflict("mapl")
> setenv("MAPL_ROOT","/lfs/h2/emc/eib/save/hang.lei/forgdit/nco_wcoss2/install2/intel-19.1.3.304/cray-mpich-8.1.12/mapl/2.53.0-esmf-8.8.0")
> whatis("Name: mapl")
> whatis("Version: 2.53.0-esmf-8.8.0")
> whatis("Category: library")
> whatis("Description: MAPL is a foundation layer of the GEOS architecture")
> […](#)

Regression test passed on WCOSS2 (Cactus) using esmf/8.8.0 and mapl/2.53.0-esmf-8.8.0 modules installed in this location.

**Comment by Hang-Lei-NOAA** [comment]
My runs used Dusan’s branch also succeeded in selected sensitive regression
tests.

It shows that we finally successfully solved the Esmf-mapl-gocart issue in
Dusan’s branch. Great job.

**Comment by ulmononian** [comment]
@DusanJovic-NOAA just curious what your ufs_common.lua lib versions were for these runs, or if you have a branch i could check out? (going to test w/ this esmf+mapl pair w/ spack-stack/1.9.0 on ursa).

**Comment by DusanJovic-NOAA** [comment]
> [@DusanJovic-NOAA](https://github.com/DusanJovic-NOAA) just curious what your ufs_common.lua lib versions were for these runs, or if you have a branch i could check out? (going to test w/ this esmf+mapl pair w/ spack-stack/1.9.0 on ursa).

I used this branch for testing: https://github.com/DusanJovic-NOAA/ufs-weather-model/tree/esmf880_mapl2530

**Comment by lipan-NOAA** [comment]
The ESMF8.8.0 and MAPL 2.53.0 libraries installed by Hang Lei have been tested on Cactus and run successfully in conjunction with the latest GOCART versions from the NASA repository. It may be necessary to update the GOCART repository in the UFS after the official releases of ESMF8.8.0 and MAPL 2.53.0 become available.

**Comment by junwang-noaa** [comment]
MAPL lib 2.53.4 was committed to UFS WM in SS 1.9.2 PR #2650. The issue is closed.
